src/twisted/internet/defer.py

```diff
@@ -295,60 +362,80 @@ def __init__(self, canceller=None):
         @type canceller: a 1-argument callable which takes a L{Deferred}. The
             return result is ignored.
         """
-        self.callbacks = []
+        self.callbacks = []  # type: List[_Callbacks]
         self._canceller = canceller
         if self.debug:
             self._debugInfo = DebugInfo()
             self._debugInfo.creator = traceback.format_stack()[:-1]
 
     def addCallbacks(
-        self,
-        callback,
-        errback=None,
-        callbackArgs=None,
-        callbackKeywords=None,
-        errbackArgs=None,
-        errbackKeywords=None,
-    ):
+        self: _DeferredT,
+        callback: DeferredCallback,
+        errback: DeferredErrback = passthru,
+        callbackArgs: _CallbackOrderedArguments = (),
+        callbackKeywords: _CallbackKeywordArguments = _NONE_KWARGS,
+        errbackArgs: _CallbackOrderedArguments = (),
+        errbackKeywords: _CallbackKeywordArguments = _NONE_KWARGS,
+    ) -> _DeferredT:
         """
         Add a pair of callbacks (success and error) to this L{Deferred}.
 
         These will be executed when the 'master' callback is run.
 
         @return: C{self}.
-        @rtype: a L{Deferred}
         """
+        # Default value used to be None and callers may be using None
+        if errback is None:
+            errback = passthru  # type: ignore[unreachable]
```

Comment:
Why make this (incorrectly) unreachable rather than making the arguments properly `Optional`?

Reply:
To be clear I'm probably in favor of making this signature change later, `None`s are a bummer, but it seems like the first iteration of the type signature should be more descriptive of the existing behavior.

Reply:
(One reason to make this change later is that if there's going to be this kind of runtime/type-check-time distinction, we should add some deprecation warnings to the runtime cases as well so that we can get the ball rolling on the removal, and that's a bunch of additional code.)

Reply:
I guess I was assuming that there weren't many callers passing in `None` versus not passing in the argument at all, and if they did so… a type warning wouldn't be too onerous.

I agree that a runtime issue would require a deprecation and that should be another PR.

But I see a type error as a "very soft" deprecation that you can ignore your way past until such time as you get a formal deprecation.

If you feel that's a bridge too far, I'll make these optional, but as a client, I wouldn't mind cleaning this up as part of better type checking.

Reply:
I think I am committed to having no further opinion on this.  Mainly I think it's unnecessary churn which would also cause valid clients adopting type-checking for the first time to have to rewrite stuff in the process (particularly it's annoying to have to replace a `None` with a `lambda x:x` or an `import passthru`) but also I think this is an edge case, and the *main* problem is that it provokes too much discussion about the pros & cons, so I'll cut that short and say "do what you like with it, either option is OK" :).

Reply:
Unless there's a consensus otherwise, I think it's better to clean up the type here as long as it stays compatible. `lambda x:x` is a minor drag for `errback`, sure, but the other values are `()` and `{}`… it just seems easy enough to deal with.

Reply:
We can also have a change of heart later if it's causing headaches.